File: Linked List/singly.cpp
#include <iostream>

template <class T>
class SinglyLinkedList {
private:
    struct Node {
        T data;
        Node* next;
        Node(T data) {
            this->data = data;
            next = nullptr;
        };
    };
    Node* root;
public:
    SinglyLinkedList() {
        root = nullptr;
    };
    void insert(T data) {
        Node* newNode = new Node(data);
        if (root == nullptr) {
            root = newNode;
            return;
        };
        Node* temp = root;
        while (temp->next != nullptr)
            temp = temp->next;
        temp->next = newNode;
    };
    void remove(T data) {
        if (root == nullptr)
            return;
        Node* temp = root;
        if (temp->data == data) {
            root = root->next;
            delete temp;
        };
        while (temp != nullptr) {
            if (temp->next->data == data) {
                Node* newTemp = temp->next;
                temp->next = temp->next->next;
                delete newTemp;
                return;
            };
            temp = temp->next;
        };
    };
    bool search(T data) {
        if (root == nullptr)
            return false;
        Node* temp = root;
        while (temp != nullptr) {
            if (temp->data == data)
                return true;
            temp = temp->next;
        };
        return false;
    };
    void reverse() {
        if (root == nullptr || root->next == nullptr)
            return;
        Node* prev = nullptr;
        Node* curr = root;
        Node* next = nullptr;
        while (curr != nullptr) {
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        };
        root = prev;
    };
    void print() {
        if (root == nullptr)
            return;
        Node* temp = root;
        while (temp != nullptr) {
            std::cout << temp->data << " ";
            temp = temp->next;
        }
        std::cout << std::endl;
    };
};

int main() {
    SinglyLinkedList<int> l;
    for (int i = 0; i < 10; i++)
        l.insert(i);
    l.reverse();
    l.print();
    return 0;
};

File: Linked List/singly.go
package main

import "fmt"

type Node[T comparable] struct {
	Data T
	Next *Node[T]
}

func NewNode[T comparable](data T) *Node[T] {
	return &Node[T]{Data: data, Next: nil}
}

type SinglyLinkedList[T comparable] struct {
	Root *Node[T]
}

func (s *SinglyLinkedList[T]) insert(data T) {
	node := NewNode(data)
	if s.Root == nil {
		s.Root = node
		return
	}
	temp := s.Root
	for {
		if temp.Next == nil {
			break
		}
		temp = temp.Next
	}
	temp.Next = node
}

func (s *SinglyLinkedList[T]) remove(data T) {
	if s.Root == nil {
		return
	}
	if s.Root.Data == data {
		s.Root = s.Root.Next
	}
	temp := s.Root
	for {
		if temp.Next.Data == data {
			temp.Next = temp.Next.Next
			return
		}
		temp = temp.Next
	}
}

func (s *SinglyLinkedList[T]) search(data T) bool {
	if s.Root == nil {
		return false
	}
	temp := s.Root
	for {
		if temp == nil {
			break
		}
		if temp.Data == data {
			return true
		}
		temp = temp.Next
	}
	return false
}

func (s *SinglyLinkedList[T]) reverse() {
	if s.Root == nil || s.Root.Next == nil {
		return
	}
	var prev, curr, next *Node[T]
	prev = nil
	curr = s.Root
	next = nil
	for {
		if curr == nil {
			break
		}
		next = curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}
	s.Root = prev
}

func (s *SinglyLinkedList[T]) print() {
	if s.Root == nil {
		return
	}
	temp := s.Root
	for {
		if temp == nil {
			break
		}
		fmt.Print(temp.Data, " ")
		temp = temp.Next
	}
}

func main() {
	var l SinglyLinkedList[int]
	for i := 0; i < 10; i++ {
		l.insert(i)
	}
	l.reverse()
	l.print()
	fmt.Println("")
}
